src/Kriti/Parser.hs

```diff
@@ -137,6 +145,32 @@ stringLit = match \case
   Lex.StringLit s -> Just s
   _               -> Nothing
 
+stringTem :: Parser [Either Text Lex.TokenStream]
+stringTem = do
+  res <- match \case
+    Lex.StringTem s -> Just $ traverse (traverse Lex.lexer) $ splitText s
+    _               -> Nothing
+  case res of
+    Left err  -> P.customFailure err
+    Right tem -> pure tem
+
+splitText :: Text -> [Either Text Text]
+splitText t = reverse $ go t []
+  where
+    go str acc =
+      if T.null str
+        then acc
+        else
+          let (txt, rest) = T.breakOn "${" str
+          in if T.null rest || T.length rest == 2
+            then Left str : acc
+            else
+              case T.findIndex (== '}') (T.drop 2 rest) of
+                Nothing -> Left str : acc
+                Just i ->
+                  let (tok, rest') = T.splitAt i (T.drop 2 rest)
+                  in go (T.drop 1 rest') (Right tok : Left txt : acc)
+
```

Comment:
Using `PatternGuards` or `MultiWayif` might prevent this from drifting so far to the right:

```suggestion
    go str acc
      | T.null str = acc
      | let (txt, rest) = T.breakOn "${" str
      , T.null rest || T.length rest == 2
      = Left str : acc
      | Just i <- T.findIndex (== '}') (T.drop 2 rest)
      = let (tok, rest') = T.splitAt i (T.drop 2 rest)
          in go (T.drop 1 rest') (Right tok : Left txt : acc)
      | otherwise = Left str : acc
```

Reply:
But also, why isn't this written using parser combinators, or something similar? Presumably it also needs to handle escapes?

Reply:
I totally blanked on the fact that `let` is just an expression and can be put in guard. : )

I wrote my reason for not using a parsig library in another comment:

> If we were to use Megaparsec to do this we would need an entirely seperate ParsecT type for consuming Text values vs our main one which uses our custom TokenStream. I believe this would lead to a complicated situation with regestering errors from this sub parser with our main parser.

However as I think about it now, this doesn't seem like a huge deal to do.

Basically we would have a second parsing type:
```
type ParserTxt = Parsec Void Text
```

Then we would run that parser on the `StringTem` to break up the sub expressions. Then we would lex the results and run in our main `Parser` as is currently done. If there are errors in `ParserTxt` we wouldn't be able to register them directly with our main `Parser`. However, we could extend the custom error type (currently just `LexError`) to also include errors from `ParserTxt` and then throw them in the main `Parser`.

I'll see if I can get this working without too much trouble today. If it goes quickly then ill commit it to this branch.

Reply:
I opted to add an extensive comment to `splitText` explaining the options and why I didn't go with parser combinators. We can easily change the implementation in the future but at the moment I don't think it buys us much.